homeassistant/components/siren/__init__.py

```diff
@@ -0,0 +1,134 @@
+"""Component to interface with various sirens/chimes."""
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    SERVICE_TOGGLE,
+    SERVICE_TURN_OFF,
+    SERVICE_TURN_ON,
+    STATE_ON,
+)
+from homeassistant.core import ServiceCall
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.config_validation import (  # noqa: F401
+    PLATFORM_SCHEMA,
+    PLATFORM_SCHEMA_BASE,
+)
+from homeassistant.helpers.entity import ToggleEntity
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+from homeassistant.loader import bind_hass
+
+from .const import (
+    ATTR_AVAILABLE_TONES,
+    ATTR_DURATION,
+    ATTR_TONE,
+    ATTR_VOLUME_LEVEL,
+    DOMAIN,
+    SUPPORT_DURATION,
+    SUPPORT_TONES,
+    SUPPORT_TURN_OFF,
+    SUPPORT_TURN_ON,
+    SUPPORT_VOLUME_SET,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(seconds=60)
+
+TURN_ON_SCHEMA = {
+    vol.Optional(ATTR_TONE): cv.string,
+    vol.Optional(ATTR_DURATION): cv.positive_float,
+    vol.Optional(ATTR_VOLUME_LEVEL): cv.small_float,
+}
+
+
+@bind_hass
+def is_on(hass: HomeAssistantType, entity_id: str) -> bool:
+    """
+    Return if the siren is on based on the state machine.
+
+    Async friendly.
+    """
+    return hass.states.is_state(entity_id, STATE_ON)
+
+
+def filter_turn_on_params(siren: SirenEntity, params: dict[str, Any]) -> dict[str, Any]:
```

Comment:
Should we use a typed dict for params?

It is hard to introduce into existing entity components, but this is new.

Reply:
I keep getting the following error, is it because I am using a constant?

```
homeassistant/components/siren/__init__.py:77: error: Expected TypedDict key to be string literal  [misc]
homeassistant/components/siren/__init__.py:79: error: Expected TypedDict key to be string literal  [misc]
homeassistant/components/siren/__init__.py:81: error: Expected TypedDict key to be string literal  [misc]
```

here's the offending code, it's the pops that are doing it:
```
if not supported_features & SUPPORT_TONES:
        params.pop(ATTR_TONE, None)
    if not supported_features & SUPPORT_DURATION:
        params.pop(ATTR_DURATION, None)
    if not supported_features & SUPPORT_VOLUME_SET:
        params.pop(ATTR_VOLUME_LEVEL, None)
```

Reply:
OK I was missing that I needed to name the TypedDict parameters instead of referencing the constants. But now I am getting the following:
```homeassistant/components/siren/__init__.py:77: error: Key "tone" of TypedDict "SirenTurnOnServiceParametersDataType" cannot be deleted  [misc]```

Reply:
...nevermind it appears I forgot how to use TypedDict's